drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,206 @@
+        Ok(self.read_flash_status())
+    }
+
+    fn bulk_erase(
+        &mut self,
+        _: &RecvMessage,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if !matches!(protect, HfProtectMode::AllowModificationsToSector0) {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        todo!()
+    }
+
+    fn page_program(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+        data: LenLimit<Leased<R, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr < SECTOR_SIZE_BYTES
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_write(
+            addr,
+            &mut LeaseBufReader::<_, 32>::from(data.into_inner()),
+        )
+        .map_err(|()| RequestError::went_away())
+    }
+
+    /// Erases the 64 KiB sector containing the given address
+    ///
+    /// If this is sector 0, requires `protect` to be
+    /// `HfProtectMode::AllowModificationsToSector0` (otherwise this function
+    /// will return an error).
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        offset: u32,
+        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        self.flash_read(
+            offset,
+            &mut LeaseBufWriter::<_, 32>::from(dest.into_inner()),
+        )
+        .map_err(|_| RequestError::went_away())
+    }
+
+    fn sector_erase(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr < SECTOR_SIZE_BYTES
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_sector_erase(addr);
+        Ok(())
+    }
+
+    fn get_mux(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfMuxState, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_mux(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfMuxState,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn get_dev(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfDevSelect, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_dev(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfDevSelect,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn hash(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        len: u32,
+    ) -> Result<[u8; SHA256_SZ], RequestError<HfError>> {
+        let hash_driver = drv_hash_api::Hash::from(HASH.get_task_id());
+        if hash_driver.init_sha256().is_err() {
+            return Err(HfError::HashError.into());
```

Comment:
should failures to initialize the hash driver also be ringbuf'd, so that they can be distinguished from `Trace::HashUpdateError`?

Reply:
Sure, done in f239a0a7b7cc